vault-plugins/shared/memdb/memdb.go

```diff
@@ -0,0 +1,407 @@
+	rels := append(t.schema.CascadeDeletes[table], t.schema.CheckingRelations[table]...) //nolint:gocritic
+	return t.processRelations(rels, obj, t.checkRelationShouldBeEmpty, ErrNotEmptyRelation)
+}
+
+func (t *Txn) checkCheckingRelations(table string, obj interface{}) error {
+	rels := t.schema.CheckingRelations[table]
+	return t.processRelations(rels, obj, t.checkRelationShouldBeEmpty, ErrNotEmptyRelation)
+}
+
+// implement main loop checking relations
+func (t *Txn) processRelations(relations []Relation, obj interface{},
+	relationHandler func(originObjectFieldValue interface{}, key Relation) error, relationHandlerError error) error {
+	valueIface := reflect.ValueOf(obj)
+	if valueIface.Type().Kind() != reflect.Ptr {
+		return fmt.Errorf("obj `%s` is not ptr", valueIface.Type())
+	}
+	msgs := []string{}
+	for _, key := range relations {
+		field := valueIface.Elem().FieldByName(key.OriginalDataTypeFieldName)
+		if !field.IsValid() {
+			return fmt.Errorf("obj `%s` does not have the field `%s`", valueIface.Type(), key.OriginalDataTypeFieldName)
+		}
+		checkedFieldValue := field.Interface()
+		if err := relationHandler(checkedFieldValue, key); err != nil {
+			msgs = append(msgs, err.Error())
+		}
+	}
+	builder := strings.Builder{}
+	if len(msgs) != 0 {
+		for _, msg := range msgs {
+			if builder.Len() != 0 {
+				builder.WriteString(";")
+			}
+			builder.WriteString(msg)
+		}
+		return fmt.Errorf(builder.String()+":%w", relationHandlerError)
+	}
+	return nil
+}
+
+func (t *Txn) checkRelationShouldBeEmpty(checkedFieldValue interface{}, key Relation) error {
+	relatedRecord, err := t.First(key.RelatedDataType, key.RelatedDataTypeFieldIndexName, checkedFieldValue)
+	if err != nil {
+		return fmt.Errorf("getting related record:%w", err)
+	}
+	if relatedRecord != nil {
+		return fmt.Errorf("relation should be empty: %q found at table %q by index %q",
+			checkedFieldValue, key.RelatedDataType, key.RelatedDataTypeFieldIndexName)
+	}
+	return nil
+}
+
+func (t *Txn) deleteChildren(parentObjectFiledValue interface{}, key Relation) error {
+	if key.indexIsSliceFieldIndex {
+		return nil
+	}
+	iter, err := t.Get(key.RelatedDataType, key.RelatedDataTypeFieldIndexName, parentObjectFiledValue)
+	if err != nil {
+		return fmt.Errorf("getting related record:%w", err)
+	}
+	for {
+		relatedRecord := iter.Next()
+		if relatedRecord == nil {
+			break
+		}
+		relatedRecord, err := t.First(key.RelatedDataType, key.RelatedDataTypeFieldIndexName, parentObjectFiledValue)
+		if err != nil {
+			return fmt.Errorf("getting related record:%w", err)
+		}
+		err = t.CascadeDelete(key.RelatedDataType, relatedRecord)
+		if err != nil {
+			return fmt.Errorf("deleting related record: at table %q by index %q, value %q",
+				key.RelatedDataType, key.RelatedDataTypeFieldIndexName, parentObjectFiledValue)
+		}
+	}
+	return nil
+}
+
+func (t *Txn) archiveChildren(archivingTimeStamp int64, archivingHash int64) func(originObjectFieldValue interface{}, key Relation) error {
```

Comment:
``` 
archivingTimeStamp int64, archivingHash int64
```

This dudes always come together. I'd make a dedicated struct for this pair. 

```go
type ArchiveMark struct {
  TimeStamp, Hash int64
}
```

Reply:
TODO